src/Utils/AST.php

```diff
@@ -608,4 +610,36 @@ public static function getOperation(DocumentNode $document, $operationName = nul
 
         return false;
     }
+
+    /**
+     * Returns the operation given a document and operation name.
+     *
+     * If a name is not provided, an operation is only returned if only one is
+     * provided in the document.
+     *
+     * @api
+     */
+    public static function getOperationAST(DocumentNode $document, ?string $operationName = null) : ?OperationDefinitionNode
+    {
+        $operation = null;
+        foreach ($document->definitions->getIterator() as $node) {
+            if (! $node instanceof OperationDefinitionNode) {
+                continue;
+            }
+
+            if ($operationName === null) {
+                // If no operation name was provided, only return an Operation if there
+                // is one defined in the document. Upon encountering the second, return
+                // null.
+                if ($operation !== null) {
+                    return null;
```

Comment:
I would consider throwing an Exception here.

Reply:
Distinguishing this case with an exception doesn't make much sense I think since it mostly increases the calling site's burden without providing a better error then the other times that null gets returned.

This is the calling logic that I use in a server-like implementation:
```
            // If no operation was found we can't continue.
            $operation = AST::getOperationAST($document, $msg->getOperationName());
            if ($operation === NULL) {
              throw new OperationNotFoundException('Could not identify operation.');
            }
```

The [current usage in this library](https://github.com/webonyx/graphql-php/blob/521efa8f089e865404afc339c56d1e1077f92e10/src/Server/Helper.php#L292) is as follows and when switching over to the new (undeprecated) method would also be unnecessarily complicated by moving to an exception instead of sticking to the reference implementation.

```
           $operationType = AST::getOperation($doc, $op->operation);

            if ($operationType === false) {
                throw new RequestError('Failed to determine operation type');
            }

            if ($operationType !== 'query' && $op->isReadOnly()) {
                throw new RequestError('GET supports only query operation');
            }
            ```

Regarding your comment
> This special case is a somewhat hidden implementation detail, I believe it does not lead to robust usage.
The [formal specification](https://spec.graphql.org/June2018/#sec-Lone-Anonymous-Operation) for lone anonymous operations actually doesn't allow multiple anonymous operations, and covers this case. 

So it's not just an implementation detail but spec compliant. Although the reporting to a client is left for other functions (which is mirrored in the examples above).

Given multiple anonymous operations (and no named operations) then the set of `operations` would be bigger then `1` thus `anonymous` must be empty, which means this is an invalid document.

Reply:
There is a substantial amount of duplication for functionality that is basically:

```
(DocumentNode $document, ?string $operationName): OperationDefinitionNode | null | Error
```

https://github.com/webonyx/graphql-php/blob/521efa8f089e865404afc339c56d1e1077f92e10/src/Experimental/Executor/Collector.php#L72-L102

https://github.com/webonyx/graphql-php/blob/521efa8f089e865404afc339c56d1e1077f92e10/src/Executor/ReferenceExecutor.php#L140-L167

...and now this implementation.

Reply:
Ah I see 🤔 The js implementation contains [the same duplication](https://github.com/graphql/graphql-js/blob/d6e760cf7060217bd25fa934bd104ff2400aad96/src/execution/execute.js#L312). It's only made more noticeable in graphql-php since it contains two Executor implementations with the logic.

A few weeks ago I would've been all for cleaning up the PHP code sites, but now that I'm at my second contribution I'm actually leaning towards sticking to the duplication that's present in the graphql-js library (and if need be, updating there before here). Since it makes development a lot easier since all the JS examples out there suddenly map nicely to PHP, even for more complex things.